### DISCUSSION TITLE
Same model with/without deepspeed different results

### DISCUSSION THREAD

**Discussion by BernardoPalmer** [deepspeedai/DeepSpeed#7322] [Q&A]
Maybe someone can enlighten me if im missing something.
Im using a single gpu RTX 4060 8gb vram to train a model, my intention with deepspeed is to use cpu offload and eventually ssd/nvme offload. I used my model without deepspeed and got for the top categories about 90% accuracy, then with deepspeed I wanted to test if I would get something similar using the exact same model architecture and hyper parameters but the results were considerably worse.
deepspeed version 0.6.19
torch 2.6.0
cuda 12
ubuntu 24

ds_zero3_cpu.json:
{
  "train_batch_size":            128,     
  "gradient_accumulation_steps": 16,   
  "optimizer": {
    "type": "AdamW",          
    "params": {
      "lr": 1e-4,
      "betas": [0.9, 0.999],
      "eps": 1e-8,
      "weight_decay": 0.01
    }
  },   
  "fp16": { "enabled": true },
  "zero_optimization": {
    "stage": 3,
    "offload_param":     { "device": "cpu", "pin_memory": true },
    "offload_optimizer": { "device": "cpu", "pin_memory": true },
    "overlap_comm":      false,
    "contiguous_gradients": true
  }
}

running with: deepspeed train.py   --deepspeed   --deepspeed_config ds_zero3_cpu.json

please let me know if more context is needed

**Comment by alanhuangyoo** [comment]
To put a number on the premise that ZeRO should not change your results: I measured this recently
for unrelated reasons. Same seed, same data order, same optimizer, 2 ranks, comparing final
parameters against a `stage: 0` baseline:

```
stage 1                                   max param diff 0.000e+00
stage 2                                   max param diff 0.000e+00
stage 3                                   max param diff 0.000e+00
stage 2 + overlap_comm + contiguous_gradients   0.000e+00
stage 2 + reduce_scatter: false                 0.000e+00
stage 3 + overlap_comm                          0.000e+00
stage 3 + stage3_param_persistence_threshold    0.000e+00
reduce_bucket_size: 1  /  stage3_prefetch_bucket_size: 0   0.000e+00
gradient_accumulation_steps: 2 (stage 2 vs 3)   0.000e+00
gradient_clipping: 1.0 (stage 2 vs 3)           0.000e+00
```

Bit-identical, not just close. So the partitioning is not where a 90% → "considerably worse" gap
comes from, and it is worth looking at the parts of the config that *do* change the math.

Three candidates in yours, in the order I would check them:

**1. `"fp16": {"enabled": true}`.** This is the big one. Enabling it here puts the whole model in
fp16 with dynamic loss scaling, which is not the same as what your non-DeepSpeed run was almost
certainly doing (fp32, or `torch.autocast` which keeps master weights in fp32). On a small model
that difference alone can cost real accuracy. Try `"fp16": {"enabled": false}` first — if you are
memory constrained, `"bf16": {"enabled": true}` is far more forgiving than fp16 on a 4060.

**2. The batch arithmetic.** `train_batch_size: 128` with `gradient_accumulation_steps: 16` on one
GPU means a micro-batch of 8 — DeepSpeed derives
`train_batch_size = micro_batch × grad_accum × world_size`. If your baseline used a batch of 128
per step, you are now taking 1/16 as many optimizer steps per epoch at the same LR. Same
hyperparameters on paper, different training run. Log `engine.train_micro_batch_size_per_gpu()`
and your step count and compare against the baseline.

**3. Who owns the LR schedule.** If your baseline had a warmup/decay schedule driven by your own
loop and DeepSpeed is not configured with the same one, the LR trajectory differs even though
`lr: 1e-4` matches in both.

Also worth checking: `deepspeed 0.6.19` with `torch 2.6.0` is an unusual pairing — 0.6.x predates
torch 2 entirely. If that version string is accurate rather than a typo for 0.16.x, upgrading is
worth doing before debugging anything else.

Quickest way to isolate it: run `stage: 0` with the same DeepSpeed config, no offload, fp16 off.
That keeps DeepSpeed in the loop but removes partitioning, precision and offload from the
picture. If that matches your baseline, the cause is one of the three above rather than ZeRO.
